web/pdf_find_controller.js

```diff
@@ -331,193 +469,173 @@ class PDFFindController {
-      matchesLength.push(originalQueryLen);
+      const [matchPos, matchLen] = getOriginalIndex(
+        diffs,
+        match.index,
+        match[0].length
+      );
+
+      if (matchLen) {
+        matches.push(matchPos);
+        matchesLength.push(matchLen);
+      }
     }
     this._pageMatches[pageIndex] = matches;
     this._pageMatchesLength[pageIndex] = matchesLength;
   }
 
-  _calculateWordMatch(query, pageIndex, pageContent, pageDiffs, entireWord) {
-    const matchesWithLength = [];
-
-    // Divide the query into pieces and search for text in each piece.
-    const queryArray = query.match(/\S+/g);
-    for (let i = 0, len = queryArray.length; i < len; i++) {
-      const subquery = queryArray[i];
-      const subqueryLen = subquery.length;
+  _convertToRegExpString(query, hasDiacritics) {
+    const { matchDiacritics } = this._state;
+    let isUnicode = false;
+    query = query.replace(
+      SPECIAL_CHARS_REG_EXP,
+      (
+        match,
+        p1 /* to escape */,
+        p2 /* punctuation */,
+        p3 /* whitespaces */,
+        p4 /* diacritics */,
+        p5 /* letters */
+      ) => {
+        // We don't need to use a \s for whitespaces since the text is got with
+        // normalizeWhitespace=true, so they're all replaced by a single " ".
+
+        if (p1) {
+          // Escape characters like *+?... to not interfer with regexp syntax.
+          return `[ ]*\\${p1}[ ]*`;
+        }
+        if (p2) {
+          // Allow whitespaces around punctuation signs.
+          return `[ ]*${p2}[ ]*`;
+        }
+        if (p3) {
+          // Replace spaces by \s+ to be sure to match any spaces.
+          return "[ ]+";
+        }
+        if (matchDiacritics) {
+          return p4 || p5;
+        }
 
-      let matchIdx = -subqueryLen;
-      while (true) {
-        matchIdx = pageContent.indexOf(subquery, matchIdx + subqueryLen);
-        if (matchIdx === -1) {
-          break;
+        if (p4) {
+          // Diacritics are removed with few exceptions.
+          return DIACRITICS_EXCEPTION.has(p4.charCodeAt(0)) ? p4 : "";
         }
-        if (
-          entireWord &&
-          !this._isEntireWord(pageContent, matchIdx, subqueryLen)
-        ) {
-          continue;
+
+        // A letter has been matched and it can be followed by any diacritics
+        // in normalized text.
+        if (hasDiacritics) {
+          isUnicode = true;
+          return `${p5}\\p{M}*`;
         }
-        const originalMatchIdx = getOriginalIndex(matchIdx, pageDiffs),
-          matchEnd = matchIdx + subqueryLen - 1,
-          originalQueryLen =
-            getOriginalIndex(matchEnd, pageDiffs) - originalMatchIdx + 1;
-
-        // Other searches do not, so we store the length.
-        matchesWithLength.push({
-          match: originalMatchIdx,
-          matchLength: originalQueryLen,
-          skipped: false,
-        });
+        return p5;
       }
+    );
+
+    if (query.endsWith("[ ]*")) {
```

Comment:
Maybe add a `const trailingSpaces = "[ ]*";` (assuming that's a good name) just above this line and use that instead to avoid having to specify that string repeatedly?